optimisation/design/Model_Optimisation_Design.md

```diff
@@ -433,23 +364,158 @@ the bounding functions specified might be of the form:
+
+The total number of all `applies` entries, e.g., `sum(arrayfun(@(x)numel(obj.applies(x)),1:numel(applies)))` + `numel(datasets)*sum(cellarray(@(x)x==0,applies))`, should be equal to or greater than the number of datasets -- that is, each dataset must have at least one function which applies.
+
+### One possible implementation of `OptFunctionsArray`
+
+![possible OptFunctionsArray diagram][OptFunctionsArray_diagram]
+
+[OptFunctionsArray_diagram]: images/OptFunctionsArray_diagram.png
+
+## Optimisation Model Object
+For efficiency of model evaluation it will be worthwhile to allow for a user defined ![S(\mathbb{Q};\mathbf{p})] to be constructed from a set of multiplier, foreground, and background functions all held in `OptFunctionsArrays` objects.
+Keeping track of those `OptFunctionsArray` objects and interacting with the optimisation engine requires a new class, `OptModel`.
+
+`OptModel` should take three `OptFunctionsArray` objects
+
+- multiplier functions, ![M(a,\ldots;\mathbf{m}) \equiv \prod_i M_i(a,\ldots;\mathbf{m}_i)]
+- foreground functions, ![F(a,\ldots;\mathbf{f}) \equiv \sum_j F_j(a,\ldots;\mathbf{f}_j)]
+- background functions, ![B(a,\ldots;\mathbf{b}) \equiv \sum_k B_k(a,\ldots;\mathbf{b}_k)]
+
+[M(a,\ldots;\mathbf{m}) \equiv \prod_i M_i(a,\ldots;\mathbf{m}_i)]: https://latex.codecogs.com/svg.latex?M%28a%2C%5Cldots%3B%5Cmathbf%7Bm%7D%29%20%5Cequiv%20%5Cprod_i%20M_i%28a%2C%5Cldots%3B%5Cmathbf%7Bm%7D_i%29
+[F(a,\ldots;\mathbf{f}) \equiv \sum_j F_j(a,\ldots;\mathbf{f}_j)]: https://latex.codecogs.com/svg.latex?F%28a%2C%5Cldots%3B%5Cmathbf%7Bf%7D%29%20%5Cequiv%20%5Csum_j%20F_j%28a%2C%5Cldots%3B%5Cmathbf%7Bf%7D_j%29
+[B(a,\ldots;\mathbf{b}) \equiv \sum_k B_k(a,\ldots;\mathbf{b}_k)]: https://latex.codecogs.com/svg.latex?B%28a%2C%5Cldots%3B%5Cmathbf%7Bb%7D%29%20%5Cequiv%20%5Csum_k%20B_k%28a%2C%5Cldots%3B%5Cmathbf%7Bb%7D_k%29
+
+[M_i(a, \ldots; \mathbf{p}_i)]: https://latex.codecogs.com/svg.latex?M_i%28a%2C%20%5Cldots%3B%20%5Cmathbf%7Bp%7D_i%29
+[F_j(a, \ldots; \mathbf{p}_j)]: https://latex.codecogs.com/svg.latex?F_j%28a%2C%20%5Cldots%3B%20%5Cmathbf%7Bp%7D_j%29
+[B_k(a, \ldots; \mathbf{p}_k)]: https://latex.codecogs.com/svg.latex?B_k%28a%2C%20%5Cldots%3B%20%5Cmathbf%7Bp%7D_k%29
+[\mathbf{p}_i]: https://latex.codecogs.com/svg.latex?%5Cmathbf%7Bp%7D_i
+[\mathbf{p}_j]: https://latex.codecogs.com/svg.latex?%5Cmathbf%7Bp%7D_j
+[\mathbf{p}_k]: https://latex.codecogs.com/svg.latex?%5Cmathbf%7Bp%7D_k
+
+such that they collectively describe a real valued function
+
+![\rho(a,\ldots;\mathbf{p}) = M(a,\ldots;\mathbf{m}) \times F(a,\ldots;\mathbf{f}) + B(a,\ldots;\mathbf{b})]
+
+[\rho(a,\ldots;\mathbf{p}) = M(a,\ldots;\mathbf{m}) \times F(a,\ldots;\mathbf{f}) + B(a,\ldots;\mathbf{b})]: https://latex.codecogs.com/svg.latex?%5Crho%28a%2C%5Cldots%3B%5Cmathbf%7Bp%7D%29%20%3D%20M%28a%2C%5Cldots%3B%5Cmathbf%7Bm%7D%29%20%5Ctimes%20F%28a%2C%5Cldots%3B%5Cmathbf%7Bf%7D%29%20&plus;%20B%28a%2C%5Cldots%3B%5Cmathbf%7Bb%7D%29
+
```

Comment:
What about the case (just saying!) that I want to a different multiplier function for each of (or perhaps a subset of) foreground functions? E.g. where I want:

\rho(a,...; p) = \sum_j M_j(a,...; m) F_j(a,...; f) + \sum_k B_k (a,...; b)

instead of:

\rho(a,...; p) = \prod_i M_i(a,...; m) \times \sum_j F_j(a,...; f) + \sum_k B_k (a,...; b)

?

Should we cater for this use case, or should we explicitly rule it out? In principle if the foreground functions are physical models of scattering (e.g. magnons and phonons) they *should* be outputting absolute units so should not need a scaling factor between them, only a single scaling factor for the data (which may or may not have been normalised to a standard measurement).

Reply:
I think it is O.K. to require that all foreground functions that apply to a single dataset should return intensity in the same units. If a user wanted to combine two models with different units, say `mb/Sr/meV/f.u.` vs. `mb/Sr/meV/atom` they would need to wrap one model in a simple function that applies the appropriate conversion factor.

If you disagree with my opinion then we can add the possibility of per-foreground-scale-factors in one of two ways

- a more general construction of &rho;(a,...;p) at the expense of extra bookkeeping
- providing *all* functions with an overall scale-factor as part of `OptFunction` that could be a single value (defaulting to 1) or another `OptFunction`

Reply:
I agree - just wanted to raise the point